### ISSUE TITLE
TracingCommandGateway should deliver CommandMessage to parent send() method

### ISSUE THREAD

**Issue by Sam-Kruglov** [AxonFramework/extension-tracing#14]
In `send` you call `GenericCommandMessage#asCommandMessage` and use the result to build the span tags, including message id. But you actually send just the payload in `super#send`, so the command message is created again somewhere up the chain.

I don't know what is the best solution here. I am writing my own gateway, maybe if I get somewhere nice I will submit a PR

**Comment by Sam-Kruglov** [comment]
got an idea, maybe just put that code into a handler interceptor and call `span#setTag`

**Comment by smcvb** [comment]
I am pretty sure this is just a bug and that we should push through the `CommandMessage` to `super#send(Object)` instead of the `commandPayload`. Thanks for filing this issue @Sam-Kruglov!

Additionally, this would be great as contribution, if you're up for the task! ;-)

**Comment by Sam-Kruglov** [comment]
Don't know if changing the superclass is a good idea, that would not be very clean. The interceptor works fine in my code. I am planning to publish the sample code very soon and we'll discuss it some more

**Comment by smcvb** [comment]
I didn't mean to adjust the `DefaultCommandGateway` (read: the 'super') implementation, but to change the method you talked about to pass the created `CommandMessage` which is set to variable `cmd` in the `super#send` call.
So, changing line 84 from:
- `super.send(command, (CommandCallback<C, R>) (commandMessage, commandResultMessage) -> {`
to:
- `super.send(cmd, (CommandCallback<C, R>) (commandMessage, commandResultMessage) -> {`

**Comment by Sam-Kruglov** [comment]
@smcvb I think if you do that, the `AbstractCommandGateway#send` will just wrap it again lol

**Comment by smcvb** [comment]
@Sam-Kruglov that is a correct assumption indeed! But I should've shed some light why calling that function again isn't an issue.

If you look in to the `GenericCommandMessage#asCommandMessage(Object)` method starting on line 47, you can see that the function performs a check if it is already dealing with an implementation of the `CommandMessage`. If that's the case, it'll just pass it along.

That thus mean that even though the`GenericCommandMessage#asCommandMessage(Object)` is called twice, the second iteration will not cause any harm in the form of 'double-wrapping'.

I could've been clearer about this, so sorry for the unspecific explanation on top.
Regardless, I hope this clarifies my train of thought!

**Comment by smcvb** [comment]
So...after a lot of label switching (aka thinking) I've decided to use the wrapped message in the `TracingCommandGateway#send(C, CommandCallback<? super C, ? super R>)` method.

Firstly, because as mentioned above we verify if the given `command` is already wrapped in a `CommandMessage`, thus resolving duplicate wrapping.
Secondly, because some information, like the `messageId`, was being reported on the `Span`.

By using the original `command` on the `delegate`, we unintentionally reported the wrong message id for `CommandMessage` upon dispatching. This problem has been alleviated in commit 1d5655b, hence why this issue has been closed.
